src/is_master.rs

```diff
@@ -34,72 +43,166 @@ pub(crate) fn is_master_command(api: Option<&ServerApi>) -> Command {
         }
         .into());
     }
+
+    if let Some(ref handler) = handler {
+        let event = ServerHeartbeatStartedEvent {
+            server_address: conn.address.clone(),
+        };
+        handler.handle_server_heartbeat_started_event(event);
+    }
+
     let start_time = Instant::now();
-    let response = conn.send_command(command, None).await?;
+    let response_result = conn.send_command(command, None).await;
     let end_time = Instant::now();
 
+    let round_trip_time = end_time.duration_since(start_time);
+
+    match parse_response_result(response_result, round_trip_time) {
+        Ok(is_master_reply) => {
+            if let Some(ref handler) = handler {
+                let event = ServerHeartbeatSucceededEvent {
+                    duration: round_trip_time,
+                    reply: is_master_reply.command_response.clone(),
+                    server_address: conn.address.clone(),
+                };
+                handler.handle_server_heartbeat_succeeded_event(event);
+            }
+            Ok(is_master_reply)
+        }
+        Err(err) => {
+            if let Some(ref handler) = handler {
+                let event = ServerHeartbeatFailedEvent {
+                    duration: round_trip_time,
+                    failure: err.clone(),
+                    server_address: conn.address.clone(),
+                };
+                handler.handle_server_heartbeat_failed_event(event);
+            }
+            Err(err)
+        }
+    }
+}
+
+fn parse_response_result(
+    response_result: Result<RawCommandResponse>,
+    round_trip_time: Duration,
+) -> Result<HelloReply> {
+    let response = response_result?;
     let server_address = response.source_address().clone();
     let basic_response = response.into_document_response()?;
     basic_response.validate()?;
     let cluster_time = basic_response.cluster_time().cloned();
-    let command_response: IsMasterCommandResponse = basic_response.body()?;
+    let command_response: HelloCommandResponse = basic_response.body()?;
 
-    Ok(IsMasterReply {
+    Ok(HelloReply {
         server_address,
         command_response,
-        round_trip_time: Some(end_time.duration_since(start_time)),
+        round_trip_time: Some(round_trip_time),
         cluster_time,
     })
 }
 
 #[derive(Debug, Clone)]
-pub(crate) struct IsMasterReply {
+pub(crate) struct HelloReply {
     pub server_address: ServerAddress,
-    pub command_response: IsMasterCommandResponse,
+    pub command_response: HelloCommandResponse,
     pub round_trip_time: Option<Duration>,
     pub cluster_time: Option<ClusterTime>,
 }
 
+/// The response to a `hello` command.
+///
+/// See the documentation [here](https://docs.mongodb.com/manual/reference/command/hello/) for more details.
 #[derive(Debug, Clone, Default, Deserialize)]
 #[serde(rename_all = "camelCase")]
-pub(crate) struct IsMasterCommandResponse {
+#[non_exhaustive]
+pub struct HelloCommandResponse {
```

Comment:
So this type is currently in a bit of a weird place in that some fields are optional where they shouldn't be (see RUST-726). Also, it's used quite a bit internally for handshakes and what not, so it has some risk of needing to be duplicated if we ever need to change it. Given those concerns and that SDAM monitoring is mostly a logging / debugging tool, I think we may want to keep this internal to the driver and just serve the reply as a `Document`. This will also allow the driver to log fields that it doesn't know about (e.g. from newer server versions). What do you think?